src/qoi/src/spectroscopic_absorption.C

```diff
@@ -0,0 +1,74 @@
+//-----------------------------------------------------------------------bl-
+//--------------------------------------------------------------------------
+//
+// GRINS - General Reacting Incompressible Navier-Stokes
+//
+// Copyright (C) 2014-2016 Paul T. Bauman, Roy H. Stogner
+// Copyright (C) 2010-2013 The PECOS Development Team
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the Version 2.1 GNU Lesser General
+// Public License as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc. 51 Franklin Street, Fifth Floor,
+// Boston, MA  02110-1301  USA
+//
+//-----------------------------------------------------------------------el-
+
+
+// This class
+#include "grins/spectroscopic_absorption.h"
+#include "grins/absorption_coeff.h"
+#include "grins/integrated_function.h"
+
+// GRINS
+#include "grins/multiphysics_sys.h"
+#include "grins/assembly_context.h"
+#include "grins/materials_parsing.h"
+#include "grins/single_variable.h"
+
+// libMesh
+#include "libmesh/getpot.h"
+#include "libmesh/fem_system.h"
+#include "libmesh/quadrature.h"
+#include "libmesh/fem_function_base.h"
+
+namespace GRINS
+{
+  SpectroscopicAbsorption::SpectroscopicAbsorption( const std::string & qoi_name, SharedPtr<libMesh::FEMFunctionBase<libMesh::Real> > absorb, SharedPtr<RayfireMesh> rayfire)
+    : IntegratedFunction<libMesh::FEMFunctionBase<libMesh::Real> >(2 /* QGauss order */,absorb,rayfire,qoi_name)
+  {}
+
+  QoIBase * SpectroscopicAbsorption::clone() const
+  {
+    return new SpectroscopicAbsorption( *this );
+  }
+
+  void SpectroscopicAbsorption::parallel_op( const libMesh::Parallel::Communicator & communicator,
+                                            libMesh::Number & sys_qoi,
+	                                        libMesh::Number & local_qoi )
+  {
+    QoIBase::parallel_op(communicator,sys_qoi,local_qoi);
+
+    // absorbance is calculated in [cm^-1], but path length is given in [m]
+    // 100.0 factor converts pathlength to [cm]
+    QoIBase::_qoi_value = std::exp( -QoIBase::_qoi_value * 100.0 );
```

Comment:
I think I figured out how this works, but I wanted to double check. For my tests, it seems that `thread_join` is called, and then `parallel_op`. Is this always the case? Because then I can just leave `thread_join` alone to do a sum, and use `parallel_op` to do the exponentiation.